### PR TITLE
azurerm_iothub - support for the `endpoint.resource_group` property

### PR DESCRIPTION
**Pull request by neil-yechenwei**
fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/3295

--- PASS: TestAccAzureRMIotHub_standard (310.27s)
--- PASS: TestAccAzureRMIotHub_ipFilterRules (321.22s)
--- PASS: TestAccAzureRMIotHub_fallbackRoute (321.66s)
--- PASS: TestAccAzureRMIotHub_requiresImport (343.44s)
--- PASS: TestAccAzureRMIotHub_fileUpload (375.62s)
--- PASS: TestAccAzureRMIotHub_basic (409.36s)
--- PASS: TestAccAzureRMIotHub_customRoutes (463.73s)
--- PASS: TestAccAzureRMIotHub_withDifferentEndpointResourceGroup (582.51s)
PASS

### CODE REVIEW

File: azurerm/internal/services/iothub/iothub_resource.go [hunk 1]
```diff
@@ -239,30 +240,37 @@ func resourceArmIotHub() *schema.Resource {
 							},
 							Sensitive: true,
 						},
+
+						"endpoint_resource_group_name": azure.SchemaResourceGroupNameForDataSource(),
```
Comment (katbyte) [inline azurerm/internal/services/iothub/iothub_resource.go:244]: I don't think we need to prefex this with endpoint as it is already in the endpoint block?

Comment (katbyte) [reply azurerm/internal/services/iothub/iothub_resource.go:244]: and this is a breaking change as it makes this property required? could we not used shared schema and write an optional one by hand?

File: website/docs/r/iothub.html.markdown [hunk 2]
```diff
@@ -157,6 +157,8 @@ An `endpoint` block supports the following:
 
 * `connection_string` - (Required) The connection string for the endpoint.
 
+* `endpoint_resource_group_name` - (Required) The name of the resource group of the endpoint.
```
Comment (katbyte) [inline website/docs/r/iothub.html.markdown:160]:
```suggestion
* `endpoint_resource_group_name` - (Required) The resource group in which the endpoint will be created.
```

File: azurerm/internal/services/iothub/tests/iothub_resource_test.go [hunk 3]
```diff
@@ -503,3 +524,91 @@ resource "azurerm_iothub" "test" {
...
 }
+
+func testAccAzureRMIotHub_withDifferentEndpointResourceGroup(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
Comment (katbyte) [inline azurerm/internal/services/iothub/tests/iothub_resource_test.go:535]:
```suggestion
  name     = "acctestRG-iothub-%d"
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @neil-yechenwei - overall this looks good but i've left some comments inline that need to be addressed before merge

File: azurerm/internal/services/iothub/iothub_resource.go (hunk 1 above)
Comment (neil-yechenwei) [reply azurerm/internal/services/iothub/iothub_resource.go:249]: updated

File: website/docs/r/iothub.html.markdown (hunk 2 above)
Comment (neil-yechenwei) [reply website/docs/r/iothub.html.markdown:162]: updated

File: azurerm/internal/services/iothub/tests/iothub_resource_test.go (hunk 3 above)
Comment (neil-yechenwei) [reply azurerm/internal/services/iothub/tests/iothub_resource_test.go:733]: updatd

**Comment by neil-yechenwei** [comment]
@katbyte , thanks for your comments. I've updated code. Please have a look. Thanks.

**Comment by neil-yechenwei** [comment]
@katbyte , thanks for your comments. I've updated code. Please have a look. Thanks.

File: azurerm/helpers/azure/resource_group.go [hunk 4]
```diff
@@ -61,12 +61,12 @@ func SchemaResourceGroupNameSetOptional() *schema.Schema {
...
 		Elem: &schema.Schema{
 			Type:         schema.TypeString,
-			ValidateFunc: validateResourceGroupName,
+			ValidateFunc: ValidateResourceGroupName,
 		},
 	}
 }
 
-func validateResourceGroupName(v interface{}, k string) (warnings []string, errors []error) {
+func ValidateResourceGroupName(v interface{}, k string) (warnings []string, errors []error) {
```
Comment (katbyte) [inline azurerm/helpers/azure/resource_group.go:69]: rather then exporting, could we just add a new schema function `SchemaResourceGroupNameOptional` and use that instead?

**Review by katbyte — APPROVED** [review]
Thanks @neil-yechenwei, aside from one comment that needs to be addressed this LGTM 👍

File: azurerm/helpers/azure/resource_group.go (hunk 4 above)
Comment (neil-yechenwei) [reply azurerm/helpers/azure/resource_group.go]: updated

**Comment by neil-yechenwei** [comment]
@katbyte , thanks for your comment. I've updated code by your comment. Please have a look. Thanks.

**Review by katbyte — APPROVED** [review]
Thanks @neil-yechenwei! this LGTM now

**Pull request merged**